### DISCUSSION TITLE
Pandora support in Music Assistant

### DISCUSSION THREAD

**Discussion by tplas** [music-assistant/support#393] [Feature requests and ideas]
Support for Pandora as a music provider in Music Assistant would be appreciated. Looks like Pandora has pretty extensive API set. I'm OK w requiring Pandora paid subscription (just like current Spotify Premium requirement).

**Comment by erkr** [comment]
Thanks Tim!

**Reply by Snobound2** [reply]
Any progress on Pandora?

**Reply by OzGav** [reply]
We have no dev working on this at the moment. Always looking for more help!

**Reply by tplas** [reply]
Unfortunately the last time I did any app dev was in the previous millennium.  :)

But thanks for the reply; I do understand.

**Comment by Snobound2** [comment]
Thanks for the update.  I wouldn't be much help unless you need someone
with MS BASIC skills.

**Comment by Traxtar928** [comment]
Also would love to get this integrated into music assistant!

**Comment by JimBrandon1240** [comment]
Any update on this...

**Reply by OzGav** [reply]
No sorry. Still need someone to volunteer to code up the provider.

**Comment by PhillySports** [comment]
I would love to get Pandora integrated into Music Assistant.  Pandora, in my opinion, is still the best service for just turning it on and letting it play (if you are looking for a few hours of great music).  Pandora has the best mix of songs you know/like and "music discovery" based upon those songs.

**Reply by PhillySports** [reply]
I would offer any help - but I don't really know anything about it.

**Comment by rfep2016** [comment]
Is there someone willing to point me in the direction of where to start a Pandora integration?  I currently have some time to put into a project and am willing to look into it, but I'm completely new to Home Assistant and just stumbled into the Music Assistant. Pandora being my preferred streaming interface, I feel I should make an attempt!

**Reply by OzGav** [reply]
Come to discord and post in the music players dev channel

**Comment by gpend** [comment]
Please vote in the poll!!
https://github.com/orgs/music-assistant/discussions/2356

**Comment by mattmaas** [comment]
Bump

**Comment by jazzmonger** [comment]
So pandora support is possible, just that it hasn't been implemented yet? It won the poll at 26% over all the other missing services....

**Reply by OzGav** [reply]
It’s a long list of possibles with a short list of developers unfortunately

**Comment by harrisonhjones** [comment]
Do we have an idea of how difficult this would be to do? I don't have a huge amount of time but I'd love to get this working and if I knew the time investment that would help me.

**Reply by OzGav** [reply]
It depends on how comfortable you are with Python and how much time you have to devote. You can look at the codebase to see examples of other providers. Help can be sought on Discord

**Reply by jazzmonger** [reply]
If we're ever going to ditch Alexa in favor of the new crop of voice assistants (like Future Proof Homes), Pandora support will be key.  I'd hate having to keep a renamed Alexa just for that one function.

**Comment by DanGarion** [comment]
This is the one missing feature that MA still needs and in turn HA. All these other player options but nothing for one of the longest standing online music systems... I've tied Pianobar so many times and always failed to get it working with HA. I used to have LMS working with Pandora but it is iffy most the time with my setup. But I know if the MA team does it, this will be rock solid.

**Comment by harrisonhjones** [comment]
It sounds like a good first step is getting Pandora to work with a Python library? Looks like https://github.com/mcrute/pydora might work for our use-case? I need to do a bit more digging.

**Comment by PhillySports** [comment]
Pandora has a developer API link.  But it is restricted to "partner
developers" which sounds like there is probably a cost involved.
Unfortunately, not an area I know much about or I'd be happy to try to help.

https://developer.pandora.com/docs/key-concepts/apis/

**Comment by DanGarion** [comment]
This dev was able to get it working with LMS https://github.com/sspiff/lms-plugin-pyrrha maybe there is something there that can be used to help figure it out.

Which of course was built off of some of this work with the original LMS https://github.com/LMS-Community/slimserver/tree/public/8.4/Slim/Plugin/Pandora

**Comment by PhillySports** [comment]
Really excited to see this get some attention! Again, I don't have any
experience but would be happy to help with testing or anything else where I
may be of use!

**Comment by jazzmonger** [comment]
Any progress on Pandora support? It won the poll and is in position #1....

**Reply by OzGav** [reply]
No. Too many other priorities at the moment.

**Reply by harrisonhjones** [reply]
@OzGav looks like the first step would be to reach out and see if Pandora would grant y'all partner status: https://developer.pandora.com/docs/getting-started/becoming-pandora-partner/ but it doesn't look like they are accepting new applications right now. Does anyone know anyone that might work for Pandora and help inquire if they'd be open to partnering with Home Assistant (Nabu Casa probably) or Music Assistant?

**Reply by OzGav** [reply]
Thanks for the info but the very first step is to find a dev who wants to do this work.

**Reply by harrisonhjones** [reply]
Oh, 100%. Let's say I'm that dev. Would y'all mind me reaching out on y'all's behalf?

**Reply by OzGav** [reply]
We are working on the 2.4 release right now so will take a little while to get back to you.

**Comment by jazzmonger** [comment]
I just contacted CHRIS.PATALANO@GMAIL.COM. - He used to be in charge of the Pandora API a few yrs ago.   He might know who to talk to.

Get me a name, any name, at Pandora and I'll find their phone number and email address.  Don't ask how I do it.

Jeff

**Reply by harrisonhjones** [reply]
@jazzmonger any movement from Chris?

**Reply by jazzmonger** [reply]
Not a peep. Give me any other names you want to chase and I will get you phone numbers and email addresses, both business and personal. It's insane we don't have this integration yet.

**Reply by jazzmonger** [reply]
Anyone?  its **crazy** that Pandora isnt in MA yet.... based on the latest poll, it is the #1 requested feature missing.

**Reply by OzGav** [reply]
No new devs are able to invest the time unfortunately...

**Reply by jazzmonger** [reply]
How much time would it realistically take?

**Reply by OzGav** [reply]
Someone would have to look at the library and the MA example provider code to answer that

**Reply by jamesleroy03** [reply]
I want Pandora also. Thank you.

**Reply by cybergrimes** [reply]
> Get me a name, any name, at Pandora and I'll find their phone number and email address. Don't ask how I do it.
> 

There’s many Pandora employees on LinkedIn, this is the only guy who mentioned any kind of API work 🤷‍♂️

https://www.linkedin.com/mwlite/profile/in/javahaiwang

**Reply by jazzmonger** [reply]
waiting on return info...

**Comment by fireheadman** [comment]
+1 for Pandora

**Comment by corradomatt** [comment]
What's the best way to get setup for development here?  Can someone point me to docs about local setup for MA addon development please?  I might have time to work on this soon.  I use a mac for my professional work but can use windows if needed.

**Reply by OzGav** [reply]
https://github.com/music-assistant/server/blob/dev/DEVELOPMENT.md

**Reply by OzGav** [reply]
And you can ask for help here https://discord.com/channels/753947050995089438/1260489193763639338

**Comment by Excha0** [comment]
Just voicing my support for Pandora inclusion, but it appears that all access to Pandora's API is gated behind a developer application, and that they aren't accepting any new developer applications right now. Otherwise, I've got some time on my hands and would love to help out with this.

In the meantime, while this is in no way ideal, I added an input_toggle to my HomeAssistant named "Pandora Playing" and a routine to an Echo dot that monitors this exposed sensor and plays Pandora when it is "on" and stops when it is "off". Adding a button to my HA instance that manipulates this sensor was easy. This gives me an easy way to stop and start Pandora from HA. It does not provide any way to manipulate what is playing, but Pandora is very very good about picking stuff, so I almost never adjust any of that anyway. Also, of course, there's not much control over where the sound comes out, other than what's connected to your Echo dot.

**Comment by cmmh** [comment]
Slimserver/Logitech Music Server / Lyrion Music Server got this implemented in a few few weeks by an awesome user when mySB was shutting down.  Surely some work here would be re-usable?  https://github.com/sspiff/lms-plugin-pyrrha

**Reply by tplas** [reply]
That would be great. And that is indeed all the functionality I would be looking for, from this integration--Just fetch playlists & play them. No need for acct mgmt, station curation, etc.  I'd prefer to do that stuff elsewhere; I just want to play Pandora thru Music Assistant.

**Comment by Hedda** [comment]
FYI, [Mopidy](https://mopidy.com/) (a other music server) has an extension written in Python for playing music from [Pandora](https://www.pandora.com/) which code could maybe be ported to MA?

- https://github.com/mopidy/mopidy-pandora

  - https://pypi.org/project/Mopidy-Pandora/

> "_Mopidy-Pandora backend extension - A backend for playing music from the [Pandora](https://www.pandora.com/) music streaming and radio service._"

  - https://mopidy.com/ext/pandora/

That in turn uses the pydora library (Python Pandora API Client) from mcrute:

* https://github.com/mcrute/pydora

  * https://pypi.org/project/pydora

Dependencies:

* Requires a Pandora user account. Users with a Pandora Premium subscription will have access to the higher quality 192 Kbps audio stream. Free accounts will play advertisements.

* pydora >= 2,<3. The Python Pandora API Client. Pre-requisites: As of 1.11.0 pydora requires [cryptography](https://pypi.python.org/pypi/cryptography). Depending on your setup, there may be some native dependencies on openssl that you will need to install beforehand. See the [cryptography installation docs](https://cryptography.io/en/latest/installation/) for details.

* cachetools >= 1.0.0. Extensible memoizing collections and decorators. The package is available as cachetools on PyPI.

* Mopidy >= 3.0.0. The music server that Mopidy-Pandora extends.

* requests >= 2.5.0. Python HTTP Requests for Humans™.

PS: Note to those wanting to try this extension in Mopidy server; "_Mopidy needs [dynamic playlists](https://github.com/mopidy/mopidy/issues/620) and [core extensions](https://github.com/mopidy/mopidy/issues/1100) to properly support Pandora. In the meantime, Mopidy-Pandora comes bundled with a frontend extension that automatically adds more tracks to the tracklist as needed. Mopidy-Pandora will ensure that there are always just two tracks in the tracklist: the currently playing track and the track that is up next. It is not possible to mix Pandora and non-Pandora tracks for playback at the same time, so any non-Pandora tracks will be removed from the tracklist when playback starts. Pandora expects users to interact with tracks at the point in time and in the sequence that it serves them up. For this reason, trying to save tracks to playlists or messing with the Mopidy-Pandora generated tracklist is probably not a good idea. And not recommended._"

**Comment by jazzmonger** [comment]
That certainly sounds promising.

**Comment by jazzmonger** [comment]
Any progress on this??  Alexa just totally creeped me out today after talking about making pizza, 5 min later it randomly asked me if I'd like to reorder 00 flour WITHOUT using any wake word.   It's time to turn that $hit OFF.

I still cant believe some stud programmer out in the wild hasn't written a Pandora integration for MA.

I installed Mopidy on Ubuntu and got the server working but getting the Pandora extension & dependencies (pydora, cryptography, etc) working is way above my pay grade.  Also it appears that MPD clients (players) are ancient history now.

**Comment by OzGav** [comment]
A Pandora provider will be added to 2.8.0 beta 5+ which allows playback fof the radio stations from a users "collection"

**Reply by Snobound2** [reply]
Thank You! Thank You! Thank You!

**Reply by jazzmonger** [reply]
I cannot believe I just read that.  Soon I'll have cursed at Alexa for the last time!!!
A gift from the MA gods. This is GREAT news!

**Reply by mynci90** [reply]
So far the music provider is working great with all my Pandora songs and stations working thus far! Thank you @OzGav 

I had stopped keeping up with Pandora on Music/Home Assistant only to open up the settings today and see that there is a beta! I have used Pandora since 2005 when I was in middle school and I am super glad it is now works with Music Assistant especially now that the speaker synchronization via Sendspin around my home works perfectly across all the smart speakers in my home (even better than Google Home groupings which glitch!)

**Comment by PhillySports** [comment]
This is the best news yet!!!

**Reply by OzGav** [reply]
Appreciate all the positivity. Just to keep expectations you can’t search or add through the provider. What you will see is your personal radio stations and you can play those.

**Reply by jazzmonger** [reply]
We'll take it.  The Pandora app works just fine for creating stations.  
This has been  long time coming. You're gonna make a lot of families very happy. 

Currently I say "Alexa, play Sting Radio on Pandora".  'Sting Radio' being one of the stations I created in the Pandora app . I assume the functionality you're suggesting will accomplish the same thing on the Voice Assistant?

when might the 2.8.0 beta 5+ be available to test?

**Reply by jazzmonger** [reply]
when might the 2.8.0 beta 5+ be available to test?

**Reply by OzGav** [reply]
Test what? Pandora is in ~~beta 4~~. Sorry I lost track. It is available in the nightly. There is an improvement in terms of automatic reauthentication that will come soon.

**Reply by PhillySports** [reply]
> Test what? Pandora is in ~beta 4~. Sorry I lost track. It is available in the nightly. There is an improvement in terms of automatic reauthentication that will come soon.

I think by "test" he just means when will the beta (beta 4?) be pushed through Home Assistant as an update, so that those of us on the beta channel can start using it.  

And I echo all the thanks that others have shared!  Pandora support in Music Assistant will be HUGE!  Pandora - IMHO - is still by far the best provider at "music discovery" and just starting with a seed song/artist (or 2, or 3) and having it create a radio station around your selections!  Can't wait to have it in Music Assistant!

**Reply by OzGav** [reply]
Days not weeks. There is work going on with general bug fixing.

**Reply by PhillySports** [reply]
Awesome!  Not trying to rush anything, just excited to give it a whirl!

**Comment by jazzmonger** [comment]
I've been waiting on this day for YEARS.  Pandora Radio Stations in Music Assistant. Worked the first time although it's only the first 5 stations in my list, but that covers 99% of what I listen to. 
@OzGav youre a superstar. Where do you want us to log issues that we might encounter?
<img width="869" height="810" alt="image" src="https://github.com/user-attachments/assets/82dfa9ce-d6e0-4e94-be90-491a340f381e" />

**Reply by OzGav** [reply]
Where all the other issues are raised. I already have fixed expiring authentication that is waiting to be merged.

**Reply by OzGav** [reply]
I’ll look at why you are only seeing five stations. Did you look in the browse view?

**Reply by jazzmonger** [reply]
> I’ll look at why you are only seeing five stations. Did you look in the browse view?
Yes. Same 5. I have about 50 stations listed in the app.
<img width="856" height="537" alt="image" src="https://github.com/user-attachments/assets/0eecfca0-f99f-445e-9180-2732e4665c54" />

**Reply by OzGav** [reply]
Might be a limitation of the unofficial API….?

**Reply by OzGav** [reply]
I have fixed the five stations issue. I am however seeing login failures due to Pandora's bot detection. I have a cookie method ready to go but will wait and see how many people can login successfully with just username and password. Maybe I got blacklisted during testing...?

**Reply by m4rsmallow** [reply]
I updated this thread last night and got SUPER excited, installed the beta but got 403 errors when trying to login. However, tried again this morning and IT WORKS. Thank you OzGav!

**Comment by Excha0** [comment]
I had the exact same experience, except mine was like a 429 error or
something. I think there's a small PR that hasn't been merged:
https://github.com/music-assistant/server/pull/2949

That said, after waiting a bit, it works great. i'm listening to it now, in
fact. Thank you, ozgav!

**Comment by PhillySports** [comment]
Had my first chance to login this morning.  Working perfectly!

**Reply by Snobound2** [reply]
Pardon my ignorance but I can't figure out how to get the beta version loaded on my HA installation.  I would love to try the beta release.  I have a stand alone computer with HA installed.  I have not used docker (unless it is there but in the background) .  I am computer literate but mostly Windows, Linux is mostly a foreign language to me.  Maybe I just need to wait until MA includes Pandora in its regular release.   Thanks!

**Reply by OzGav** [reply]
Just go to the HA add on store, shut down the stable server and then install and start the beta server.

**Comment by Snobound2** [comment]
There it is.  I didn't realise there was a second version in the add-on
store, I just saw MA and didn't go further to see MA(beta).  I found it
now, installing.
Thank you

**Comment by acm5708** [comment]
Just installed the Beta Music Server Current version: 2.6.0b21. Not seeing Pandora as a listed Music Provider. Is there something I’m doing wrong/missing? Would LOVE to test this out, been following this thread for some time!

**Reply by acm5708** [reply]
Turns out I had an update (or 2 😜). We are up and running! I’m so excited for this!!

**Comment by detroitgoose** [comment]
I'm pretty excited for this, as I had initially held off on using Music Assistant based on Pandora being my main streaming service.

I currently control Pandora through Home Assistant via Pianobar running on a networked headless Snapcast server/client and a rather complicated nodered flow to skip tracks and change stations thanks to the Pianobar CLI commands. It allows for near-total control, including "Likes" and "dislikes". Snapcast then serves my other pi-based clients for multiroom audio.

About a year ago or so I set up a Jellyfin server and reorganized my local music files, so I've adopted MA into the mix. (That the Jellyfin player is "unmaintained" is another whole issue.)

The problem for me is that because I'm still tied to Pianobar, and MA doesn't yet offer multiple servers, I'm forced to continue using Snapcast to play both Pianobar and MA. They have to fight amongst themselves for the audio pipe.

For me, it's basically a race between `adequate Pandora support*` in MA, or SendSpin allowing multiple servers. I'd very much prefer MA and SendSpin for everything, but if I can't skip Pandora tracks in MA, I may very well stick with Pianobar and pipe it into a secondary SendSpin server when multiple servers hit the stable release. Snapcast has been very good to me for half a decade, but the many benefits to SendSpin are clear.

`*Adequate means this: Firstly, that tracks can be skipped. This is pretty crucial to me. Second, if that's not possible, that MA radio stations at least remix tracks every time they're started. Currently, in the beta, when you change between stations they start over from the same track in the same order. That's a deal-breaker for my use case. If I can't "skip" a track, I should at least be able to restart the station with a new psuedo-random order. As it stands, if I go from my "rock" station to my "jazz" station then back again, I'm confronted with songs I've already listened to.`

@OzGav, you'd have to tell me if this needs its own feature request somewhere, if there's already progress in this direction, or if remixing or skipping tracks is even possible with MA.

I know this is mostly an issue with Pandora being hostile to third party players, and I've contacted them multiple times to request that they open up their API or partnering with HA/MA devs... but they are stone cold against being as easy to implement as their competitors. It's frustrating because I otherwise prefer Panadora as a media source.

**Reply by OzGav** [reply]
Let me look at the caching I do for the track list. Skipping is more difficult

**Reply by detroitgoose** [reply]
Thanks. I see the static list persists after this made it into the stable release.

Paulus suggested opening a new issue, also expecting caching. Should I do that?

**Reply by OzGav** [reply]
No I just haven’t looked at it yet. I’ll do that soon.

**Reply by OzGav** [reply]
@detroitgoose I am getting geoblocked now which I am not easily able to get around. I have a potential fix but I would need someone to test it locally. If you are using HAOS you can install our dev server and do the following

- Stop your current server
- In Home Assistant, make sure you have 'advanced mode' toggled on in your profile settings
- From the HA App Store, install the "Music Assistant DEV SERVER" app
- In the Configuration page, fill in these details:
- for server_repo put "pandora-fresh-tracks”
- for frontend_repo put "main"
- (both without quotes)
- Hit 'Save' and start the App. This will take quite a lot longer than usual, like more than 10 mins.

Once it starts just add Pandora and try streaming to the web browser and see if after a while you can stop and then start the stream again and see if you get different tracks. Obviously check to make sure I haven’t broken anything else.

**Reply by cryptk** [reply]
Working on getting this installed now to test

**Reply by cryptk** [reply]
With the pandora-fresh-tracks branch, I'm able to start a Pandora station, wait until the third song starts, stop payback, restart the same station and it starts with a new song.

With the current 2.8.0 release if I do the same, when I restart playback, it starts with the same set of songs.

Seems like your change works!

(side note, dealing with the 429 errors from Pandora when swapping between app versions is somewhat annoying as there is currently no way to send the `{"forceActive": True}` payload to the `/station/playbackResumed` endpoint to force Pandora to stop any other streams and resume here)

**Reply by OzGav** [reply]
Thanks. I have opened a PR.

**Reply by detroitgoose** [reply]
@OzGav, I'm glad @cryptk was able to get it working, because I'm just seeing a loop of Step 1 in the logs.

(Note, I'm not too concerned with troubleshooting this error. I'm happy to test once it hits the main branch. I greatly appreciate your time with this issue! This PR will allow me to fully adopt MA and sendspin across my network.)

```
Music Assistant Developer Add-on
-----------------------------------------------------------
Building from source using server_repo: pandora-fresh-tracks
-----------------------------------------------------------
Step 1: Installing Music Assistant Server
-----------------------------------------------------------
Server repository: music-assistant/server@pandora-fresh-tracks
Server URL: git+https://github.com/music-assistant/server@pandora-fresh-tracks
Installing dependencies from: https://raw.githubusercontent.com/music-assistant/server/pandora-fresh-tracks/requirements_all.txt
error: Error while accessing remote requirements file: `https://raw.githubusercontent.com/music-assistant/server/pandora-fresh-tracks/requirements_all.txt`
```

**Reply by detroitgoose** [reply]
Track list refresh is working great in the beta app. Thanks so much @OzGav

**Comment by cryptk** [comment]
This has been working quite nicely for me! One thing I am missing though is the ability to use the higher quality audio streams. Currently the plugin is limited to the 64kbps AAC but Pandora One subscribers have access to 192kbps MP3. Being able to stream that instead would be great! This should likely be a configurable option in the provider, with fallback to the current behavior if the higher quality audio isn't available. There appears to be two ways to determine if the higher quality audio is available to a user.

The general process looks to be to check the authentication response body to determine what should be done here, config.flags appears to be a list of feature flags and if "highQualityStreamingAvailable" is in that list, then the users account is allowed to use the higher quality streams.

Additionally there is a flag in the authentication response "highQualityStreamingEnabled" which I believe is tied to the account setting for if the user wants to use the high quality audio or not (this can be set on their user profile).

I would propose that there is a new configuration option added to the provider configuration for stream quality, allowing selection between "standard" and "high". "standard" is exactly what it streams now with "high", hopefully obviously, being the higher quality audio.

If a user has selected "high" AND their account has the "highQualityStreamingAvailable" config.flag set, then the audio used by MA should be the higher quality audio. If the user selects "high" but does not have that config flag, then the "standard" audio quality should be used anyway (maybe even logging a warning saying that they requested high quality audio but their account does not appear to have a paid subscription or something).

Thoughts?

**Reply by cryptk** [reply]
@OzGav I have tested a change locally that follows the process I suggested above, namely:

- Check for the presence of the highQualityStreamingAvailable flag on the authentication object
- Add a config option to select standard or high quality audio
- If both indicate high quality should be used, request fragments for "mp3-hifi" from Pandora and also set content_type to MP3 in get_stream_details.

This seems to be working well in my testing over the last couple of days. The only catch is that because the provider fetches 1000 fragments up front, if you change the quality setting in the provider, I don't know if that would kick in until the stream is stopped and started again.

It's also not clear what would happen if the streamDetails are generated with a content type of AAC, and then the user switches to high-quality, which could potentially result in the stream details saying AAC while the audio fetched is actually MP3.

That said, it works well enough to prove out the concept that high-quality audio should be possible

**Reply by cryptk** [reply]
Submitted a PR at https://github.com/music-assistant/server/pull/3617 to implement this. I've been running this change locally for a couple of weeks so far.

**Reply by OTNSam** [reply]
Just wanted to say thank you for getting this added! Spotify deleted my podcast over Covid so I refuse to use them. 🙏

**Comment by jazzmonger** [comment]
@Ozgav, your efforts are GREATLY appreciated!

**Comment by TermeHansen** [comment]
I was waiting for the 2.8.0 release to start trying out the pandora interface of MA. Sadly I get 403 when I try to log in, and I don't understand why. I can fine log into pandora and also use pianobar with no issues...

I even made a minimal python script for just these first commands:

```
            print(LOGIN_ENDPOINT)
            print(headers)
            print(login_data)
            async with session.post(
                LOGIN_ENDPOINT,
                headers=headers,
                json=login_data,
                timeout=aiohttp.ClientTimeout(total=30)) as response:
                print(response.status)
                js = await response.json()
                print(js)

```

yielding:

```
https://www.pandora.com/api/v1/auth/login
{'Content-Type': 'application/json;charset=utf-8', 'X-CsrfToken': '2b0722efd3194154', 'Cookie': 'csrftoken=2b0722efd3194154', 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36', 'Accept': 'application/json, text/plain, */*', 'Accept-Language': 'en-US,en;q=0.9', 'Origin': 'https://www.pandora.com', 'Referer': 'https://www.pandora.com/'}
{'username': 'xxxxxx', 'password': 'xxxxxxx', 'keepLoggedIn': True, 'existingAuthToken': None}
403
{'message': 'Invalid request', 'errorCode': 1215, 'errorString': 's2s_high_score', 'appId': 'PXXljWHHUe', 'jsClientSrc': '/XljWHHUe/init.js', 'firstPartyEnabled': False, 'vid': None, 'uuid': '93613609-2aab-11f1-a941-c56ae644e533', 'hostUrl': '/XljWHHUe/xhr', 'blockScript': '/XljWHHUe/captcha?a=c&u=93613609-2aab-11f1-a941-c56ae644e533&v=&m=0&b=aHR0cDovL2xvY2FsaG9zdDo1ODExL21pbm9zL2pzb24/bWV0aG9kPWF1dGgubG9naW4=&h=UE9TVA==', 'altBlockScript': 'https://captcha.px-cloud.net/PXXljWHHUe/captcha.js?a=c&u=93613609-2aab-11f1-a941-c56ae644e533&v=&m=0&b=aHR0cDovL2xvY2FsaG9zdDo1ODExL21pbm9zL2pzb24/bWV0aG9kPWF1dGgubG9naW4=&h=UE9TVA==', 'customLogo': None}``
```

Any ideas of why this happens to me, and what I could try?

**Reply by TermeHansen** [reply]
got it working. It was due to wrong user agent. I will make a PR soon with my suggested changes...

**Comment by ccaroccio** [comment]
Idea!

When the [Next Track] button is pressed, the session is refreshed (AKA, skip track). @OzGav what do you think of the principle?

**Reply by OzGav** [reply]
Probably not as that would require hacking the core code. However, some changes have been made to the core architecture which may allow skipping but that will take time to look at unless someone else wants to give it a go.

**Reply by TermeHansen** [reply]
I might look at it...

**Reply by ccaroccio** [reply]
Currently this functionality works great on Lyrion music server, you may consider taking a look at their source code or give their community a ping

**Reply by TermeHansen** [reply]
I think I'm close to have the re-write of the code for the new dynamic playlist feature which opens up for skipping and showing of the current 4 songs of the "playlist", I will open a PR shortly and ping you @OzGav for a few general MA api usage questions for this use case...

**Comment by swamiller** [comment]
@OzGav Is it possible to add Pandora's "Shuffle" station to the station list?  I don't typically pick one station to listen to I let Pandora create a custom mix from the stations I have set up. The default apps I use (Roku and Denon primarily) have the "Shuffle" station option.

**Reply by OzGav** [reply]
It might be possible I'll try and have a look soon

**Reply by detroitgoose** [reply]
Another option is to play the "Thumbprint Radio" station that is automatically populated if you have more than 3 stations with at least 4 thumbed up tracks on each station.

The resulting station will play a mix of those song that you have given a thumbs up, as well as tracks similar to those.

I know this isn't exactly what you're looking for, but it will give you a single station that has variation from across multiple stations.

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
